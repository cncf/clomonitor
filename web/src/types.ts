export interface Organization {
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
}

export interface BaseProject {
  id: string;
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
  devstats_url?: string;
  maturity_id: Maturity;
  category_id: Category;
  score: Score;
  updated_at: number;
}

export interface Project extends BaseProject {
  repositories: BaseRepository[];
  organization: {
    name: string;
  };
}

export interface ProjectDetail extends BaseProject {
  repositories: Repository[];
}

export interface BaseRepository {
  name: string;
  url: string;
  kind: RepositoryKind;
}

export interface Repository extends BaseRepository {
  digest: string;
  repository_id: string;
  score: Score;
  reports: Report[];
}

export interface Report {
  data: CoreReport | any;
  linter_id: LinterId;
  report_id: string;
  updated_at: number;
}

export interface CoreReport {
  // ScoreType
  [key: string]: {
    [key: string]: ReportCheck;
  };
}

export interface ReportCheck {
  passed: boolean;
  value?: string;
  url?: string;
}

export type Score = {
  [key in ScoreType]: number;
} & { score_kind: ScoreKind };

export interface FiltersSection {
  name: string;
  title: string;
  filters: Filter[];
}

export interface Filter {
  name: string | number;
  label: string;
  legend?: string;
  decorator?: JSX.Element;
}

export interface Issue {
  level: number;
  description: string;
}

export interface Prefs {
  search: { limit: number; sort: { by: SortBy; direction: SortDirection } };
  theme: {
    effective: string;
  };
}

export interface ReportOptionData {
  icon: JSX.Element;
  name: string;
  legend: JSX.Element;
}

export enum Maturity {
  Graduated = 0,
  Incubating,
  Sandbox,
}

export enum Category {
  'App definition' = 0,
  Observability,
  Orchestration,
  Platform,
  Provisioning,
  Runtime,
  Serverless,
}

export enum Rating {
  A = 'a',
  B = 'b',
  C = 'c',
  D = 'd',
}

export enum FilterKind {
  Maturity = 'maturity',
  Category = 'category',
  Rating = 'rating',
}

export enum LinterId {
  core = 0,
}

export enum ScoreType {
  BestPractices = 'best_practices',
  Documentation = 'documentation',
  Global = 'global',
  Legal = 'legal',
  License = 'license',
  Security = 'security',
}

export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortBy {
  Name = 'name',
  Score = 'score',
}

export enum RepositoryKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ScoreKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ReportOption {
  Adopters = 'adopters',
  ApprovedLicense = 'approved',
  ArtifactHubBadge = 'artifacthub_badge',
  Changelog = 'changelog',
  CodeOfConduct = 'code_of_conduct',
  CommunityMeeting = 'community_meeting',
  Contributing = 'contributing',
  DCO = 'dco',
  Governance = 'governance',
  LicenseScanning = 'scanning',
  Maintainers = 'maintainers',
  OpenSSFBadge = 'openssf_badge',
  Readme = 'readme',
  RecentRelease = 'recent_release',
  Roadmap = 'roadmap',
  SecurityPolicy = 'security_policy',
  SlackPresence = 'slack_presence',
  SPDX = 'spdx_id',
  TrademarkDisclaimer = 'trademark_disclaimer',
  Website = 'website',
}

export interface SearchFiltersURL extends BasicQuery {
  pageNumber: number;
}

export interface BasicQuery {
  text?: string;
  filters?: {
    [key: string]: (string | number)[];
  };
}

export interface SearchQuery extends BasicQuery {
  limit: number;
  offset: number;
  sort_by: SortBy;
  sort_direction: SortDirection;
}

export interface SearchData {
  limit: number;
  offset: number;
  sort_by: string;
  sort_direction: string;
  text?: string;
  category?: number[];
  maturity?: number[];
  rating?: number[];
}

export interface Error {
  kind: ErrorKind;
  message?: string;
}

export enum ErrorKind {
  Other,
  NotFound,
}

export type ReportOptionInfo = {
  [key in ReportOption]: ReportOptionData;
};

export interface RecommendedTemplate {
  name: string;
  url: string;
}
