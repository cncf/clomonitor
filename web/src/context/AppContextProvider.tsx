import { isNull } from 'lodash';
import { createContext, Dispatch, useContext, useEffect, useReducer, useState } from 'react';

import { Prefs, SortBy, SortDirection } from '../types';
import getMetaTag from '../utils/getMetaTag';
import lsStorage from '../utils/localStoragePreferences';
import themeBuilder from '../utils/themeBuilder';

interface AppState {
  prefs: Prefs;
}

interface Props {
  children: JSX.Element;
}

const initialState: AppState = {
  prefs: lsStorage.getPrefs(),
};

type Action =
  | { type: 'updateTheme'; theme: string }
  | { type: 'updateLimit'; limit: number }
  | { type: 'updateSort'; by: SortBy; direction: SortDirection };

export const AppContext = createContext<{
  ctx: AppState;
  dispatch: Dispatch<any>;
}>({
  ctx: initialState,
  dispatch: () => null,
});

export function updateTheme(theme: string) {
  return { type: 'updateTheme', theme };
}

export function updateLimit(limit: number) {
  return { type: 'updateLimit', limit };
}

export function updateSort(by: SortBy, direction: SortDirection) {
  return { type: 'updateSort', by, direction };
}

export function updateActiveStyleSheet(current: string) {
  const secondary = getMetaTag('secondaryColor');
  document.getElementsByTagName('html')[0].setAttribute('data-theme', current);
  document
    .querySelector(`meta[name='theme-color']`)!
    .setAttribute('content', current === 'light' ? secondary : '#0f0e11');
}

export function appReducer(state: AppState, action: Action) {
  let prefs;
  switch (action.type) {
    case 'updateTheme':
      prefs = {
        ...state.prefs,
        theme: {
          effective: action.theme,
        },
      };

      lsStorage.setPrefs(prefs);
      updateActiveStyleSheet(action.theme);
      return {
        ...state,
        prefs: prefs,
      };

    case 'updateLimit':
      prefs = {
        ...state.prefs,
        search: {
          ...state.prefs.search,
          limit: action.limit,
        },
      };
      lsStorage.setPrefs(prefs);
      return {
        ...state,
        prefs: prefs,
      };

    case 'updateSort':
      prefs = {
        ...state.prefs,
        search: {
          ...state.prefs.search,
          sort: {
            by: action.by,
            direction: action.direction,
          },
        },
      };
      lsStorage.setPrefs(prefs);
      return {
        ...state,
        prefs: prefs,
      };

    default:
      return { ...state };
  }
}

function AppContextProvider(props: Props) {
  const activeProfilePrefs = lsStorage.getPrefs();
  const [ctx, dispatch] = useReducer(appReducer, {
    prefs: activeProfilePrefs,
  });
  const [activeInitialTheme, setActiveInitialTheme] = useState<string | null>(null);

  useEffect(() => {
    const theme = activeProfilePrefs.theme.effective || 'light';
    themeBuilder.init();
    updateActiveStyleSheet(theme);
    setActiveInitialTheme(theme);
  }, []); /* eslint-disable-line react-hooks/exhaustive-deps */

  if (isNull(activeInitialTheme)) return null;

  return <AppContext.Provider value={{ ctx, dispatch }}>{props.children}</AppContext.Provider>;
}

function useAppContext() {
  return useContext(AppContext);
}

export { AppContextProvider, useAppContext };
