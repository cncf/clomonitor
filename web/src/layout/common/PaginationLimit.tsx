import { ChangeEvent, useContext, useRef } from 'react';

import { AppContext } from '../../context/AppContextProvider';
import styles from './PaginationLimit.module.css';

const AVAILABLE_LIMITS = [20, 40, 60];

interface Props {
  onPaginationLimitChange: (limit: number) => void;
}

const PaginationLimit = (props: Props) => {
  const { ctx } = useContext(AppContext);
  const { limit } = ctx.prefs.search;
  const selectEl = useRef<HTMLSelectElement>(null);

  const handleChange = (event: ChangeEvent<HTMLSelectElement>) => {
    props.onPaginationLimitChange(parseInt(event.target.value));
    forceBlur();
  };

  const forceBlur = (): void => {
    if (selectEl && selectEl.current) {
      selectEl.current.blur();
    }
  };

  return (
    <div className="d-none d-md-flex flex-nowrap align-items-center lh-1">
      <label className="form-label me-2 mb-0">Show:</label>
      <select
        ref={selectEl}
        className={`form-select form-select-sm rounded-0 ${styles.select}`}
        value={limit}
        onChange={handleChange}
      >
        {AVAILABLE_LIMITS.map((opt: number) => {
          return (
            <option value={opt} key={`limit_${opt}`}>
              {opt}
            </option>
          );
        })}
      </select>
    </div>
  );
};

export default PaginationLimit;
