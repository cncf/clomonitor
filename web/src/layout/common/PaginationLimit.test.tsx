import { fireEvent, render, screen } from '@testing-library/react';

import { AppContext } from '../../context/AppContextProvider';
import { SortBy, SortDirection } from '../../types';
import PaginationLimit from './PaginationLimit';

const mockOnPaginationLimitChange = jest.fn();

const defaultProps = {
  onPaginationLimitChange: mockOnPaginationLimitChange,
};

const mockCtx = {
  prefs: {
    search: { limit: 20, sort: { by: SortBy.Name, direction: SortDirection.ASC } },
    theme: { effective: 'light', configured: 'light' },
  },
};

describe('PaginationLimit', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(
      <AppContext.Provider value={{ ctx: mockCtx, dispatch: jest.fn() }}>
        <PaginationLimit {...defaultProps} />
      </AppContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders proper content', () => {
    render(
      <AppContext.Provider value={{ ctx: mockCtx, dispatch: jest.fn() }}>
        <PaginationLimit {...defaultProps} />
      </AppContext.Provider>
    );

    expect(screen.getByText('Show:')).toBeInTheDocument();
    expect(screen.getByRole('combobox')).toBeInTheDocument();
    expect(screen.getAllByRole('option').length).toBe(3);
    expect(screen.getByRole('option', { name: '20' })).toBeInTheDocument();
    expect(screen.getByRole('option', { name: '40' })).toBeInTheDocument();
    expect(screen.getByRole('option', { name: '60' })).toBeInTheDocument();
    expect((screen.getByRole('option', { name: '20' }) as HTMLOptionElement).selected).toBe(true);

    expect(screen.getByLabelText('Pagination limit select')).toBeInTheDocument();
    expect(screen.getByLabelText('Pagination limit select')).toHaveValue('20');
  });

  it('renders limit from context', () => {
    render(
      <AppContext.Provider
        value={{
          ctx: { prefs: { ...mockCtx.prefs, search: { ...mockCtx.prefs.search, limit: 60 } } },
          dispatch: jest.fn(),
        }}
      >
        <PaginationLimit {...defaultProps} />
      </AppContext.Provider>
    );

    expect((screen.getByRole('option', { name: '60' }) as HTMLOptionElement).selected).toBe(true);
  });

  it('calls onChange to update select', () => {
    render(
      <AppContext.Provider value={{ ctx: mockCtx, dispatch: jest.fn() }}>
        <PaginationLimit {...defaultProps} />
      </AppContext.Provider>
    );

    const select = screen.getByRole('combobox');
    fireEvent.change(select, '40');

    expect(mockOnPaginationLimitChange).toHaveBeenCalledTimes(1);
    expect(mockOnPaginationLimitChange).toHaveBeenCalledWith(20);
  });
});
