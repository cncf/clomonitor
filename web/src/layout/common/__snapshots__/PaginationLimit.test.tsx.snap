// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationLimit creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-none d-md-flex flex-nowrap align-items-center lh-1"
  >
    <label
      class="form-label me-2 mb-0"
    >
      Show:
    </label>
    <select
      aria-label="Pagination limit select"
      class="form-select form-select-sm rounded-0 cursorPointer select"
    >
      <option
        value="20"
      >
        20
      </option>
      <option
        value="40"
      >
        40
      </option>
      <option
        value="60"
      >
        60
      </option>
    </select>
  </div>
</DocumentFragment>
`;
